server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,552 @@
+        long shardPrimaryAndCoordinatingLimits = tracker.getPrimaryAndCoordinatingLimits();
+        boolean shardMemoryLimitsBreached = ((double)shardCombinedBytes / shardPrimaryAndCoordinatingLimits) > this.upperOperatingFactor;
+
+        if(shardMemoryLimitsBreached) {
+            // Secondary Parameters (i.e. LastSuccessfulRequestDuration and Throughput) is taken into consideration when
+            // the current node utilization is greater than primary_parameter.node.soft_limit of total node limits.
+            if(((double)nodeTotalBytes / this.shardIndexingPressureSettings.getNodePrimaryAndCoordinatingLimits()) < this.nodeSoftLimit) {
+                boolean isShardLimitsIncreased = this.increaseShardPrimaryAndCoordinatingLimits(tracker);
+                if(isShardLimitsIncreased == false) {
+                    tracker.getPrimaryOperationTracker().getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                    totalNodeLimitsBreachedRejections.incrementAndGet();
+                }
+                return !isShardLimitsIncreased;
+            } else {
+                boolean shardLastSuccessfulRequestDurationLimitsBreached =
+                    this.evaluateLastSuccessfulRequestDurationLimitsBreached(tracker.getPrimaryOperationTracker().getPerformanceTracker(),
+                        requestStartTime);
+
+                if(shardLastSuccessfulRequestDurationLimitsBreached) {
+                    tracker.getPrimaryOperationTracker().getRejectionTracker()
+                        .incrementLastSuccessfulRequestLimitsBreachedRejections();
+                    totalLastSuccessfulRequestLimitsBreachedRejections.incrementAndGet();
+                    return true;
+                }
+
+                boolean shardThroughputDegradationLimitsBreached =
+                    this.evaluateThroughputDegradationLimitsBreached(tracker.getPrimaryOperationTracker().getPerformanceTracker(),
+                        tracker.getPrimaryOperationTracker().getStatsTracker(),
+                        primaryAndCoordinatingThroughputDegradationLimits);
+
+                if (shardThroughputDegradationLimitsBreached) {
+                    tracker.getPrimaryOperationTracker().getRejectionTracker()
+                        .incrementThroughputDegradationLimitsBreachedRejections();
+                    totalThroughputDegradationLimitsBreachedRejections.incrementAndGet();
+                    return true;
+                }
+
+                boolean isShardLimitsIncreased = this.increaseShardPrimaryAndCoordinatingLimits(tracker);
+                if(isShardLimitsIncreased == false) {
+                    tracker.getPrimaryOperationTracker().getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                    totalNodeLimitsBreachedRejections.incrementAndGet();
+                }
+
+                return !isShardLimitsIncreased;
+            }
```

Comment:
Redundant logic, Lets break it down as 
```
public void onShardMemoryLimitIncreased(OperationTracker opTracker, Predicate<OperationTracker> isShardLimitsIncreasedPredicate, double degradationLimits, double shardLimits) {
   blah....
}
```

Reply:
Agree, this was planned and was added as a ToDo in the Draft PR description. Has now been taken care of in the final PR.